File: app/src/main/java/com/zionhuang/music/viewmodels/SongsViewModel.java
package com.zionhuang.music.viewmodels;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;

import com.zionhuang.music.db.SongEntity;
import com.zionhuang.music.repository.SongRepository;

import java.util.List;

import io.reactivex.rxjava3.disposables.CompositeDisposable;

public class SongsViewModel extends AndroidViewModel {
    private static final String TAG = "SongsViewModel";
    private CompositeDisposable compositeDisposable = new CompositeDisposable();
    private SongRepository songRepository;
    private LiveData<List<SongEntity>> allSongs;

    public SongsViewModel(@NonNull Application application) {
        super(application);
        songRepository = new SongRepository(application);
        allSongs = songRepository.getAllSongsAsLiveData();
    }

    public LiveData<List<SongEntity>> getAllSongs() {
        return allSongs;
    }

    public void insertSong(SongEntity song) {
        songRepository.insert(song);
    }

    @Override
    protected void onCleared() {
        compositeDisposable.dispose();
        super.onCleared();
    }
}

File: app/src/main/java/com/zionhuang/music/viewmodels/SongsViewModel.kt
package com.zionhuang.music.viewmodels

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.zionhuang.music.db.entities.*
import com.zionhuang.music.models.sortInfo.AlbumSortInfoPreference
import com.zionhuang.music.models.sortInfo.ArtistSortInfoPreference
import com.zionhuang.music.models.sortInfo.PlaylistSortInfoPreference
import com.zionhuang.music.models.sortInfo.SongSortInfoPreference
import com.zionhuang.music.repos.SongRepository
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn

@OptIn(ExperimentalCoroutinesApi::class)
class SongsViewModel(application: Application) : AndroidViewModel(application) {
    private val songRepository = SongRepository(application)

    val allSongsFlow = SongSortInfoPreference.flow.flatMapLatest { sortInfo ->
        songRepository.getAllSongs(sortInfo).flow
    }.stateIn(viewModelScope, SharingStarted.Lazily, emptyList())

    val allArtistsFlow = ArtistSortInfoPreference.flow.flatMapLatest { sortInfo ->
        songRepository.getAllArtists(sortInfo).flow
    }.stateIn(viewModelScope, SharingStarted.Lazily, emptyList())

    val allAlbumsFlow = AlbumSortInfoPreference.flow.flatMapLatest { sortInfo ->
        songRepository.getAllAlbums(sortInfo).flow
    }.stateIn(viewModelScope, SharingStarted.Lazily, emptyList())

    val allPlaylistsFlow = PlaylistSortInfoPreference.flow.flatMapLatest { sortInfo ->
        songRepository.getAllPlaylists(sortInfo).flow
    }.stateIn(viewModelScope, SharingStarted.Lazily, emptyList())

    val likedSongCount = songRepository.getLikedSongCount()
    val downloadedSongCount = songRepository.getDownloadedSongCount()

    fun getArtistSongsAsFlow(artistId: String) = SongSortInfoPreference.flow.flatMapLatest { sortInfo ->
        songRepository.getArtistSongs(artistId, sortInfo).flow
    }.map { list ->
        listOf(SongHeader(songRepository.getArtistSongCount(artistId), SongSortInfoPreference.currentInfo)) + list
    }

    fun getPlaylistSongsAsFlow(playlistId: String) = songRepository.getPlaylistSongs(playlistId).flow.map { list ->
        listOf(PlaylistSongHeader(list.size, list.sumOf { it.song.duration.toLong() })) + list
    }

    fun getLikedSongsAsFlow() = SongSortInfoPreference.flow.flatMapLatest { sortInfo ->
        songRepository.getLikedSongs(sortInfo).flow
    }.map { list ->
        listOf(SongHeader(list.size, SongSortInfoPreference.currentInfo)) + list
    }

    fun getDownloadedSongsAsFlow() = SongSortInfoPreference.flow.flatMapLatest { sortInfo ->
        songRepository.getDownloadedSongs(sortInfo).flow
    }.map { list ->
        listOf(SongHeader(list.size, SongSortInfoPreference.currentInfo)) + list
    }
}
